internal/providers/zvm/zvm.go

```diff
@@ -0,0 +1,130 @@
+// interface (opt/com.coreos/config).
+
+package zvm
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/coreos/ignition/v2/config/shared/errors"
+	"github.com/coreos/ignition/v2/config/v3_1_experimental/types"
+	"github.com/coreos/ignition/v2/internal/log"
+	"github.com/coreos/ignition/v2/internal/providers/util"
+	"github.com/coreos/ignition/v2/internal/resource"
+	"github.com/coreos/vcontext/report"
+)
+
+const authorizedSenders string = "*"
+
+func FetchConfig(f *resource.Fetcher) (types.Config, report.Report, error) {
+	/*fetch config files directly from reader device.*/
+	_, err := f.Logger.LogCmd(exec.Command("/sbin/modprobe", "vmur"), "loading zVM control program module")
+	if err != nil {
+		f.Logger.Err("couldn't install vmur module: %v", err)
+		return types.Config{}, report.Report{}, err
+	}
+	/* onlineDevice 000c */
+	logger := f.Logger
+	rc := onlineDevice(logger)
+	if !rc {
+		return types.Config{}, report.Report{}, err
+	}
+	f.Logger.Info("The reader device has been online.")
+	/* pullReader */
+	var comm string = "/sbin/vmur li | tail -n +2 | cut -c 1-80 | awk '{print $1,$2,$3,$10,$11}'"
+	cmd := exec.Command("/bin/sh", "-c", comm)
+	out, err := cmd.Output()
+	if err != nil {
+		f.Logger.Err("Can not get reader device: %v", err)
+	}
+	f.Logger.Info("The reader device info: ")
+	fmt.Println(string(out))
+	items := strings.Split(string(out), "\n")
+	/* Process each spool file that is class "x".*/
+	for _, records := range items {
+		record := strings.Split(records, " ")
+		f.Logger.Info("The single record info: ")
+		fmt.Println(record, len(record))
+		if len(record) == 1 {
+			break
+		}
+		originid := record[0]
+		spoolid := record[1]
+		class := record[2]
+		filename := record[3]
+		types := record[4]
+		if class != "X" {
+			/* Spool file is not of the class required for processing by this script.*/
+			//continue
+		}
+		if authorizedSenders != "*" {
+			if !strings.Contains(originid, authorizedSenders) {
+				/* Originator is not authorized to send configuration files.*/
+				//continue
+			}
+		}
+		var file string
+		if types == "" {
+			file = filename
+		} else {
+			file = filename + "." + types
+		}
+		/* Receive the spool file.*/
+		if filename == "ignition" {
+			_, err := f.Logger.LogCmd(exec.Command("/sbin/vmur", "re", "-f", spoolid, file), "Receive the spool file")
+			if err == nil {
+				f.Logger.Info("using config file at %q", file)
+				_, err := f.Logger.LogCmd(exec.Command("tar", "-xzvf", file), "Extract the spool file")
+				file = "config.ign"
+				rawConfig, err := ioutil.ReadFile(file)
+				if err != nil {
+					f.Logger.Err("couldn't read config %q: %v", filename, err)
+					break
+				}
+				return util.ParseConfig(f.Logger, rawConfig)
+			}
+		}
+		_ = file
+	}
+	return types.Config{}, report.Report{}, errors.ErrEmpty
+}
+
+func fileExists(path string) bool {
+	_, err := os.Stat(path)
+	return (err == nil)
```

Comment:
this should probably return (bool, error), and differentiate between the three cases:
1) file exists, no error returned
2) file doesn't exist, no returned
3) stat failed for a reason other than the file existing, that error returned

Reply:
Will redesign the onlineDevice function. The fileExists is not necessary ,but cio_ignore is necessary for online Device.